### PR TITLE
Upgrade to lucene-5.2.0-snapshot-1673124.

### PR DESCRIPTION
**Pull request by jpountz**

### CODE REVIEW

**Comment by mikemccand** [comment]
LGTM, I like the added checks in AllTermSpanScorer.

File: src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java [hunk 1]
```diff
@@ -51,7 +51,7 @@ public AllTermQuery(Term term) {
     }
 
     @Override
-    public Weight createWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
+    public SpanWeight createWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
```
Comment (rmuir) [inline src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java]: This class overrides weight/scorer, where is its positions check (throw IAE if proximity is not enabled) ?

File: src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java [hunk 2]
```diff
@@ -77,31 +81,43 @@ public AllTermSpanScorer scorer(LeafReaderContext context, Bits acceptDocs) thro
...
+            protected void setFreqCurrentDoc() throws IOException {
                 freq = 0.0f;
                 numMatches = 0;
                 payloadScore = 0;
                 payloadsSeen = 0;
-                do {
-                    int matchLength = spans.end() - spans.start();
 
-                    freq += docScorer.computeSlopFactor(matchLength);
+                assert spans.startPosition() == -1 : "incorrect initial start position, spans="+spans;
```
Comment (rmuir) [inline src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java]: I think this is fine for now, but note this stuff got excessive in all the spans and still would not find bugs. In order to make things debuggable i had to solve it another way: https://issues.apache.org/jira/browse/LUCENE-6411

So I am not sure about all the asserts, to me I just get lost in the code.

**Comment by rmuir** [comment]
I actually disagree with mike. Paul went the route of adding these asserts to every single spans, and yet they could still not be debugged. 

Problem is: when trying to duplicate checks like this across N classes then its inevitable that some will be incomplete. you are also bound to have some buggy asserts (false fails) and it also makes it really hard to read the code.

So I think the asserts should be removed (from lucene too). I opened https://issues.apache.org/jira/browse/LUCENE-6411 yesterday and worked on it all day long to fix the root cause.

File: src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java (hunk 2 above)
Comment (jpountz) [inline src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java]: I did not want to do anything smart, this just copies over logic from SpanScorer.

Comment (rmuir) [inline src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java]: Why do we need the copy-paste at all? This whole thing seems like a code duplication of PayloadTermQuery.

File: src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java (hunk 1 above)
Comment (jpountz) [inline src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java]: we need to fix it in Lucene too then

File: src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java (hunk 2 above)
Comment (rmuir) [inline src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java]: And i just did not have the time to yet yesterday remove the stupid asserts from SpanScorer. Please, lets not drag this stuff in again. If oyu want to push fine, but you will see a second push from me removing all this crap.

File: src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java (hunk 1 above)
Comment (rmuir) [inline src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java]: Maybe you are a bit behind. I added tests for this position check last night and fixed bugs in the default impl.

**Comment by jpountz** [comment]
While I do agree that AssertingSpans is a better way to find bugs, all that this patch does is aligning AllTermQuery with SpanTermQuery from which it is derived. I can push a new update once LUCENE-6411 is done.

**Comment by rmuir** [comment]
LUCENE-6411 is already done. Please don't add these redundant asserts. it just causes me more work to remove them.

File: src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java (hunk 2 above)
Comment (jpountz) [inline src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java]: The duplication comes from the fact that AllTermScorer needs to process the payload in the middle of the setFreqCurrentDoc loop.

**Comment by jpountz** [comment]
I can remove them but then you might want to remove them from SpanScorer.setFreqCurrentDoc too.

File: src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java (hunk 2 above)
Comment (rmuir) [inline src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java]: That is just what PayloadTermQuery does, too. So again how is this different from PayloadTermScorer? Even the variable/method names are similar.

**Comment by rmuir** [comment]
The only reason they are still there is because i needed to get some amount of sleep last night. It is work to cleanup all this stuff.

**Comment by rmuir** [comment]
See my commit, i cleaned up this query: it doesn't manipulate Spans anymore.

**Comment by jpountz** [comment]
Thanks @rmuir !

**Pull request merged**